src/nsrt_learning/side_predicate_learning.py

```diff
@@ -394,53 +397,73 @@ def _sideline(self) -> List[PartialNSRTAndDatastore]:
-                    # and we are making them more specific, it is guaranteed
-                    # that such a grounding exists, otherwise our 1:1
-                    # option:operator assumption is violated.
-                    new_pnad = self._try_specializing_pnad(
-                        necessary_add_effects, pnad, segment)
-                    assert new_pnad is not None, "1:1 assumption violated"
-                    assert new_pnad.option_spec == pnad.option_spec
+                # We start by checking if any of the PNADs associated with the
+                # demonstrated option are able to match this transition.
+                for pnad in pnads_for_option:
+                    ground_op = self._find_unification(necessary_add_effects,
+                                                       pnad, segment)
+                    if ground_op is not None:
+                        obj_to_var = dict(
+                            zip(ground_op.objects, pnad.op.parameters))
+                        if len(param_opt_to_nec_pnads[option.parent]) == 0:
+                            pnad_op_name = pnad.op.name + "0"
+                            pnad.op = pnad.op.copy_with(name=pnad_op_name)
+                            param_opt_to_nec_pnads[option.parent].append(pnad)
+                        break
+                # If we weren't able to find a substitution (i.e, the above
+                # for loop did not break), we need to try
+                # specializing each of our PNADs.
+                else:
+                    for pnad in pnads_for_option:
+                        new_pnad = self._try_specializing_pnad(
+                            necessary_add_effects, pnad, segment)
+                        if new_pnad is not None:
+                            assert new_pnad.option_spec == pnad.option_spec
+                            if len(param_opt_to_nec_pnads[option.parent]) > 0:
+                                param_opt_to_nec_pnads[option.parent].remove(
+                                    pnad)
+                            del pnad
+                            break
+                    # If we were unable to specialize any of the PNADs, we need
+                    # to split from the most general PNAD and make a new PNAD
+                    # to cover these necessary add effects.
+                    else:
+                        new_pnad = self._try_specializing_pnad(
+                            necessary_add_effects,
+                            param_opt_to_general_pnad[option.parent], segment)
+                        assert new_pnad is not None
+
+                    # Add the new PNAD to the dictionary mapping options to
+                    # PNADs. Note that even if there are more than 10
+                    # PNADs associated with an option, the below check will
+                    # still work.
+                    if not new_pnad.op.name[-1].isdigit():
```

Comment:
i'm confused about the interaction between this line and the place above where you add "0" to the end. when does this line get hit?

also is there any hope that we can purge all of this naming stuff and just rename the operators once at the end? i think that would be cleaner, if it doesn't break anything

Reply:
good idea on just naming things at the end! I think we should be able to do this; lemme try it!